### PR TITLE
Upgrading to TS 2.4

### PR DESCRIPTION
**Pull request by iclanton**

### CODE REVIEW

File: apps/rush/src/utilities/test/PackageChangeAnalyzer.test.ts [hunk 1]
```diff
@@ -31,7 +31,7 @@ describe('PackageChangeAnalyzer', () => {
       }
     };
 
-    PackageChangeAnalyzer.getPackageDeps = (path: string, ignored: string[]) => repoHashDeps;
+    PackageChangeAnalyzer.getPackageDeps = (packagePath: string, ignored: string[]) => repoHashDeps;
```
Comment (qz2017) [inline apps/rush/src/utilities/test/PackageChangeAnalyzer.test.ts]: >packagePath [](start = 44, length = 11)

What is this for?

Comment (iclanton) [inline apps/rush/src/utilities/test/PackageChangeAnalyzer.test.ts]: This violated the `no-shadowed-variable` TSLint rule.

---
In reply to: [129388208](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129388208) [](ancestors = 129388208)

File: apps/rush/src/utilities/test/PublishUtilities.test.ts [hunk 2]
```diff
@@ -150,9 +150,11 @@ describe('sortChangeRequests', () => {
 
 describe('isRangeDependency', () => {
   it('can test ranges', () => {
+    /* tslint:disable:no-unused-expression */
     expect(PublishUtilities.isRangeDependency('>=1.0.0 <2.0.0')).is.true;
     expect(PublishUtilities.isRangeDependency('1.0.0')).is.false;
     expect(PublishUtilities.isRangeDependency('^1.0.0')).is.false;
     expect(PublishUtilities.isRangeDependency('~1.0.0')).is.false;
+    /* tslint:enable:no-unused-expression */
```
Comment (nickpape-msft) [inline apps/rush/src/utilities/test/PublishUtilities.test.ts]: >/* tslint:enable:no-unused-expression */ [](start = 4, length = 40)

This is another reason I prefer to use the assert syntax

File: libraries/api-extractor/src/ApiDefinitionReference.ts [hunk 3]
```diff
@@ -14,7 +16,7 @@
  * \@microsoft/sp-core-library:Guid.equals
  * es6-collections:Map
  */
-export interface IApiDefinintionReferenceParts {
+export interface IApiDefinitionReferenceParts {
```
Comment (qz2017) [inline libraries/api-extractor/src/ApiDefinitionReference.ts]: >IApiDefinitionReferenceParts [](start = 17, length = 28)

Seriously!

Comment (iclanton) [inline libraries/api-extractor/src/ApiDefinitionReference.ts]: Yeah "Definition" was misspelled.

---
In reply to: [129390968](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129390968) [](ancestors = 129390968)

File: apps/rush/src/utilities/test/PublishUtilities.test.ts (hunk 2 above)
Comment (iclanton) [inline apps/rush/src/utilities/test/PublishUtilities.test.ts]: Yeah it's sorta annoying.

---
In reply to: [129390566](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129390566) [](ancestors = 129390566)

File: core-build/gulp-core-build-typescript/src/defaultTslint.json [hunk 4]
```diff
@@ -13,9 +13,25 @@
...
+          "private-instance-field",
+          "public-static-method",
+          "protected-static-method",
+          "private-static-method",
+          "public-constructor",
+          "public-instance-method",
+          "protected-constructor",
+          "protected-instance-method",
+          "private-constructor",
+          "private-instance-method"
```
Comment (qz2017) [inline core-build/gulp-core-build-typescript/src/defaultTslint.json]: Where does this order come from?

Comment (iclanton) [inline core-build/gulp-core-build-typescript/src/defaultTslint.json]: It's as similar to our existing order as possible.

---
In reply to: [129392775](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129392775) [](ancestors = 129392775)

File: apps/rush/tslint.json [hunk 5]
```diff
@@ -50,7 +50,6 @@
     "no-trailing-whitespace": true,
     "no-unnecessary-semicolons": true,
     "no-unused-expression": true,
-    "no-unused-imports": true,
```
Comment (nickpape-msft) [inline apps/rush/tslint.json]: >"no-unused-imports": true, [](start = 4, length = 26)

why?

File: apps/rush-lib/src/data/VersionPolicy.ts [hunk 6]
```diff
@@ -44,10 +44,13 @@ export abstract class VersionPolicy {
   public static load(versionPolicyJson: IVersionPolicyJson): VersionPolicy {
     const definition: VersionPolicyDefinitionName = VersionPolicyDefinitionName[versionPolicyJson.definitionName];
     if (definition === VersionPolicyDefinitionName.lockStepVersion) {
+       // tslint:disable-next-line:no-use-before-declare
       return new LockStepVersionPolicy(versionPolicyJson as ILockStepVersionJson);
     } else if (definition === VersionPolicyDefinitionName.individualVersion) {
+      // tslint:disable-next-line:no-use-before-declare
```
Comment (nickpape-msft) [inline apps/rush-lib/src/data/VersionPolicy.ts]: >// tslint:disable-next-line:no-use-before-declare [](start = 6, length = 49)

you could get rid of this by moving version policies to different files?

File: apps/rush-lib/src/index.ts [hunk 7]
```diff
@@ -47,7 +47,7 @@ export {
 
 export {
   VersionMismatchFinder
-} from './data/VersionMismatchFinder'
+} from './data/VersionMismatchFinder';
```
Comment (nickpape-msft) [inline apps/rush-lib/src/index.ts]: >VersionMismatchFinder [](start = 15, length = 21)

I wonder how this was missed?

File: common/config/rush/npm-shrinkwrap.json [hunk 8]
```diff
@@ -5,27 +5,73 @@
...
+          "version": "4.5.1",
+          "from": "tslint@>=4.5.1 <4.6.0",
+          "resolved": "https://registry.npmjs.org/tslint/-/tslint-4.5.1.tgz"
+        },
+        "tslint-microsoft-contrib": {
+          "version": "4.0.1",
+          "from": "tslint-microsoft-contrib@>=4.0.0 <4.1.0",
+          "resolved": "https://registry.npmjs.org/tslint-microsoft-contrib/-/tslint-microsoft-contrib-4.0.1.tgz"
+        },
+        "typescript": {
```
Comment (nickpape-msft) [inline common/config/rush/npm-shrinkwrap.json]: >typescript [](start = 9, length = 10)

this should not be introducing side-by-side for typescript

File: common/config/rush/npm-shrinkwrap.json [hunk 9]
```diff
@@ -5,27 +5,73 @@
...
       "version": "2.2.0",
       "from": "@microsoft/api-extractor@>=2.2.0 <3.0.0",
-      "resolved": "https://registry.npmjs.org/@microsoft/api-extractor/-/api-extractor-2.2.0.tgz"
+      "resolved": "https://registry.npmjs.org/@microsoft/api-extractor/-/api-extractor-2.2.0.tgz",
+      "dependencies": {
+        "typescript": {
+          "version": "2.2.2",
+          "from": "typescript@>=2.2.2 <2.3.0",
+          "resolved": "https://registry.npmjs.org/typescript/-/typescript-2.2.2.tgz"
+        }
```
Comment (nickpape-msft) [inline common/config/rush/npm-shrinkwrap.json]: this side-by-side will go away once new api-extractor is published

File: common/config/rush/npm-shrinkwrap.json (hunk 8 above)
Comment (iclanton) [inline common/config/rush/npm-shrinkwrap.json]: Yup. I'm going to do a followup PR as soon as this publishes.

---
In reply to: [129394536](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129394536) [](ancestors = 129394536)

File: apps/rush/tslint.json (hunk 5 above)
Comment (iclanton) [inline apps/rush/tslint.json]: It's built into TS now, so the rule has been removed.

---
In reply to: [129393749](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129393749) [](ancestors = 129393749)

File: core-build/gulp-core-build/src/tasks/CleanTask.ts [hunk 10]
```diff
@@ -1,6 +1,8 @@
 // Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
 // See LICENSE in the project root for license information.
 
+/* tslint:disable:no-trailing-whitespace whitespace */ /* Remove this when GCB-TS is published and upgraded */
```
Comment (nickpape-msft) [inline core-build/gulp-core-build/src/tasks/CleanTask.ts]: >trailing [](start = 21, length = 8)

why not just remove the whitespace?

File: apps/rush-lib/src/index.ts (hunk 7 above)
Comment (iclanton) [inline apps/rush-lib/src/index.ts]: Some of the rules missed edge cases. They're more correct now.

---
In reply to: [129394309](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129394309) [](ancestors = 129394309)

File: apps/rush-lib/src/data/VersionPolicy.ts (hunk 6 above)
Comment (iclanton) [inline apps/rush-lib/src/data/VersionPolicy.ts]: Out of scope.

---
In reply to: [129394167](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129394167) [](ancestors = 129394167)

Comment (qz2017) [inline apps/rush-lib/src/data/VersionPolicy.ts]: No, please do not change this file. It is alpha and actively under development. 

---
In reply to: [129394167](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129394167) [](ancestors = 129394167)

File: core-build/gulp-core-build/src/tasks/CleanTask.ts (hunk 10 above)
Comment (iclanton) [inline core-build/gulp-core-build/src/tasks/CleanTask.ts]: There is no whitespace - TSLint is confused with two versions of TS in the same repo. I debugged into it and couldn't find a resolution, so I'm disabling the rules temporarily until I upgrade GCB in this repo.

---
In reply to: [129395072](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129395072) [](ancestors = 129395072)

File: core-build/gulp-core-build-typescript/src/TSLintTask.ts [hunk 11]
```diff
@@ -126,24 +136,43 @@ export class TSLintTask extends GulpTask<ITSLintTaskConfig> {
...
+    let program: ts.Program | undefined = undefined;
+    try {
+      const tsconfigFileData: ITsConfigFile<Object> = TypeScriptConfiguration.getTsConfigFile(this.buildConfig);
+      const tsconfigFilePath: string =
+        path.join(
+          this.buildConfig.rootPath,
+          this.buildConfig.tempFolder,
+          'tslint-tsconfig.json'
+        );
+        fs.writeFileSync(tsconfigFilePath, JSON.stringify(tsconfigFileData, undefined, 2));
```
Comment (nickpape-msft) [inline core-build/gulp-core-build-typescript/src/TSLintTask.ts]: >fs.wri [](start = 8, length = 6)

tabs

File: libraries/api-extractor/src/DocItemLoader.ts [hunk 12]
```diff
@@ -1,11 +1,13 @@
 // Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
 // See LICENSE in the project root for license information.
 
+/* tslint:disable:no-trailing-whitespace */ /* Remove this when GCB-TS is published and upgraded */
```
Comment (nickpape-msft) [inline libraries/api-extractor/src/DocItemLoader.ts]: >/* tslint:disable:no-trailing-whitespace */  [](start = 0, length = 44)

why?

File: libraries/node-core-library/package.json [hunk 13]
```diff
@@ -25,7 +25,7 @@
     "@types/z-schema": "3.16.31",
     "fs-extra": "~0.26.7",
     "jju": "~1.3.0",
-    "typescript": "~2.2.2",
+    "typescript": "~2.4.1",
```
Comment (nickpape-msft) [inline libraries/node-core-library/package.json]: >"typescript": "~2.4.1", [](start = 4, length = 23)

I don't think this project needs a typescript dependency?

File: libraries/api-extractor/src/DocItemLoader.ts (hunk 12 above)
Comment (iclanton) [inline libraries/api-extractor/src/DocItemLoader.ts]: Same reason as your other comment.

---
In reply to: [129396045](https://github.com/Microsoft/web-build-tools/pull/276#discussion_r129396045) [](ancestors = 129396045)

**Comment by nickpape-msft** [comment]
:shipit:

**Comment by nickpape-msft** [comment]
AWS

**Comment by qz2017** [comment]
:shipit:

**Pull request merged**